geometry/cartesian_product_body.hpp

```diff
@@ -0,0 +1,80 @@
+
+#include "geometry/cartesian_product.hpp"
+
+#include <algorithm>
+
+#include "quantities/named_quantities.hpp"
+#include "quantities/tuples.hpp"
+
+namespace principia {
+namespace geometry {
+namespace internal_cartesian_product {
+
+using quantities::Apply2;
+
+template<typename LTuple, typename RTuple, int... indices>
+struct CartesianProductAdditiveGroup<LTuple, RTuple,
+                                     std::integer_sequence<int, indices...>> {
+  // The types of the result of addition and subtraction, with suitable
+  // specializations for the void case of Apply2.
```

Comment:
Should we give those semantics to quantities::Sum so that we don't need those generators?

Alternatively, would the following work?
```c++
template<typename L, typename R>
using Sum = std::conditional_t<
    std::is_void<L>{}, R,
    std::conditional_t<std::is_void<R>{}, L,
                       quantities::Sum<L, R>>>;
```

Reply:
Let's not go there.  These `void` are a bit hacky anyway.  An alternative would be to pass two extra 1-argument metafunctions `LTransform` and `RTransform` to be applied to the elements of the longer tuple.  (This might work better at the usage site as it would avoid specializations with `void`: we'd just pass the identity metafunction.)

Reply:
Done.